File: app/src/main/java/eugeen3/keepinfit/adapters/FoodItemAdapter.java
package eugeen3.keepinfit.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import eugeen3.keepinfit.R;
import eugeen3.keepinfit.entities.FoodItem;
import eugeen3.keepinfit.itemtouch.ItemTouchHelperAdapter;
import eugeen3.keepinfit.viewmodels.SharedViewModel;

import static java.lang.String.valueOf;

public class FoodItemAdapter extends RecyclerView.Adapter<FoodItemAdapter.ViewHolder> implements Filterable, ItemTouchHelperAdapter {
    private List<FoodItem> foodItems;
    private List<FoodItem> foodItemsFull;

    public void setSharedViewModel(SharedViewModel sharedViewModel) {
        this.sharedViewModel = sharedViewModel;
    }

    private SharedViewModel sharedViewModel;
    private final static String PORTION_WEIGHT = "в 100 гр:";

    @Override
    @NonNull
    public FoodItemAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.base_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(FoodItemAdapter.ViewHolder holder, int position) {
        FoodItem currentFoodItem = foodItems.get(position);
        holder.nameView.setText(currentFoodItem.getName());
        holder.massView.setText(PORTION_WEIGHT);
        holder.proteinsView.setText(valueOf(currentFoodItem.getProteins()));
        holder.carbohydratesView.setText(valueOf(currentFoodItem.getCarbohydrates()));
        holder.fatsView.setText(valueOf(currentFoodItem.getFats()));
        holder.kcalsView.setText(valueOf(currentFoodItem.getKilocalories()));
    }

    @Override
    public int getItemCount() {
        return foodItems == null ? 0 : foodItems.size();
    }

    public void setFoodItems(List<FoodItem> foodItems) {
        this.foodItems = foodItems;
        notifyDataSetChanged();
    }

    public void setFoodItemsFull(List<FoodItem> foodItems) {
        this.foodItemsFull = new ArrayList<>(foodItems);
        notifyDataSetChanged();
    }

    public FoodItem getFoodItemAt(int position) {
        return  foodItems.get(position);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private final TextView nameView, massView, proteinsView, carbohydratesView, fatsView, kcalsView;
        ViewHolder(View view){
            super(view);
            nameView = view.findViewById(R.id.tvCardTitle);
            massView = view.findViewById(R.id.tvCardSubtitle);
            proteinsView = view.findViewById(R.id.tvProteinsValue);
            carbohydratesView = view.findViewById(R.id.tvCarbohydratesValue);
            fatsView = view.findViewById(R.id.tvFatsValue);
            kcalsView = view.findViewById(R.id.tvKilocaloriesValue);
        }
    }

    @Override
    public Filter getFilter() {
        return filter;
    }

    private Filter filter = new Filter() {
        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            List <FoodItem> filteredList = new ArrayList<>();

            if (constraint == null || constraint.length() == 0) {
                filteredList.addAll(foodItemsFull);
            } else {
                String filterPattern = constraint.toString().toLowerCase().trim();

                for (FoodItem foodItem: foodItemsFull) {
                    if (foodItem.getName().toLowerCase().contains(filterPattern)) {
                        filteredList.add(foodItem);
                    }
                }
            }

            FilterResults results = new FilterResults();
            results.values = filteredList;

            return results;
        }

        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            foodItems.clear();
            foodItems.addAll((List) results.values);
            notifyDataSetChanged();
        }
    };

    @Override
    public void onItemDismiss(int position) {
        sharedViewModel.delete(foodItems.get(position));
        foodItems.remove(position);
        notifyItemRemoved(position);
    }

    @Override
    public boolean onItemMove(int fromPosition, int toPosition) {
        if (fromPosition < toPosition) {
            for (int i = fromPosition; i < toPosition; i++) {
                Collections.swap(foodItems, i, i + 1);
            }
        } else {
            for (int i = fromPosition; i > toPosition; i--) {
                Collections.swap(foodItems, i, i - 1);
            }
        }
        notifyItemMoved(fromPosition, toPosition);
        return true;
    }
}

File: app/src/main/java/eugeen3/keepinfit/adapters/FoodItemAdapter.kt
package eugeen3.keepinfit.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.ImageButton
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import eugeen3.keepinfit.R
import eugeen3.keepinfit.databinding.BaseItemBinding
import eugeen3.keepinfit.databinding.FragmentFoodItemsListBinding
import eugeen3.keepinfit.entities.FoodItem
import eugeen3.keepinfit.itemtouch.ItemTouchHelperAdapter
import eugeen3.keepinfit.ui.FoodItemFragment
import eugeen3.keepinfit.viewmodels.SharedViewModel
import java.util.*

class FoodItemAdapter : RecyclerView.Adapter<FoodItemAdapter.ViewHolder>(), Filterable, ItemTouchHelperAdapter {

    private var foodItems: MutableList<FoodItem>? = null
    private var foodItemsFull: MutableList<FoodItem>? = null
    private var sharedViewModel: SharedViewModel? = null

    inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val binding = BaseItemBinding.bind(view)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.base_item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.binding.apply {
            tvCardTitle.text = foodItems!![position].name
            tvCardSubtitle.text = PORTION_WEIGHT
            tvProteinsValue.text = foodItems!![position].proteins.toString()
            tvFatsValue.text = foodItems!![position].fats.toString()
            tvCarbohydratesValue.text = foodItems!![position].carbohydrates.toString()
            tvKilocaloriesValue.text = foodItems!![position].kilocalories.toString()
            ibEdit.setOnClickListener {
                sharedViewModel?.setFoodItem(foodItems!![position])
                TODO("open fooditemfragment")
            }
        }
    }

    override fun getItemCount(): Int {
        return foodItems?.size ?: 0
    }

    fun setSharedViewModel(sharedViewModel: SharedViewModel?) {
        this.sharedViewModel = sharedViewModel
    }

    fun setFoodItems(foodItems: List<FoodItem?>?) {
        this.foodItems = ArrayList(foodItems)
        notifyDataSetChanged()
    }

    fun setFoodItemsFull(foodItems: List<FoodItem?>?) {
        foodItemsFull = ArrayList(foodItems)
        notifyDataSetChanged()
    }

    fun getFoodItemAt(position: Int): FoodItem? {
        return foodItems!![position]
    }

    override fun getFilter(): Filter {
        return filter
    }

    private val filter: Filter = object : Filter() {
        override fun performFiltering(constraint: CharSequence): FilterResults {
            //TODO fix filters
//            val filteredList: MutableList<FoodItem> = ArrayList()
//            if (constraint.isEmpty()) {
//                filteredList.addAll(foodItemsFull as Array<FoodItem>)
//            } else {
//                val filterPattern = constraint.toString().toLowerCase(Locale.ROOT).trim { it <= ' ' }
//                for (foodItem in foodItemsFull!!) {
//                    if (foodItem.name.toLowerCase(Locale.ROOT).contains(filterPattern)) {
//                        filteredList.add(foodItem)
//                    }
//                }
//            }
//            val results = FilterResults()
//            results.values = filteredList
//            return results
            val charSearch = constraint.toString()
            if (charSearch.isEmpty()) {
                foodItems = foodItemsFull
            } else {
                val resultList = ArrayList<FoodItem>()
                for (FoodItem in foodItemsFull!!) {
                    if (FoodItem.name.toLowerCase(Locale.ROOT).contains(charSearch.toLowerCase(Locale.ROOT))) {
                        resultList.add(FoodItem)
                    }
                }
                foodItems = resultList
            }
            val filterResults = FilterResults()
            filterResults.values = foodItems
            return filterResults
        }

        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            foodItems?.clear()
            foodItems?.addAll(results?.values as Collection<FoodItem>)
            notifyDataSetChanged()
        }
    }

    override fun onItemDismiss(position: Int) {
        sharedViewModel?.delete(foodItems?.get(position))
        foodItems?.removeAt(position)
        notifyItemRemoved(position)
    }

    override fun onItemMove(fromPosition: Int, toPosition: Int): Boolean {
        if (fromPosition < toPosition) {
            for (i in fromPosition until toPosition) {
                Collections.swap(foodItems, i, i + 1)
            }
        } else {
            for (i in fromPosition downTo toPosition + 1) {
                Collections.swap(foodItems, i, i - 1)
            }
        }
        notifyItemMoved(fromPosition, toPosition)
        return true
    }

    companion object {
        private const val PORTION_WEIGHT = "в 100 гр:"
    }
}